### PR TITLE
JS-6 account restriction

### PR DESCRIPTION
**Pull request by fboucquez**
Hi @rg911 , @fullcircle23 

This PR is for the account restriction feature. The only part missing is the integration tests I will be working soon. Have a look if you have the time. 

Notes:
I need to patch the catbuffer AccountRestrictionTypeDto to allow all the possible values. I have also added a smart parser that converts a "value" in a restriction to an Address, MosaicId or TransactionType depending on the target of the restriction type.

Thanks
Fernando

### CODE REVIEW

File: sdk-core/src/main/java/io/nem/sdk/api/AccountRepository.java [hunk 1]
```diff
@@ -93,6 +94,30 @@
...
+     */
+    Observable<AccountRestrictions> getAccountRestrictions(Address address);
+
+    /**
+     * Returns the account restrictions for a given array of public keys.
+     *
+     * @param publicKeys {@link List} of {@link PublicKey}
+     * @return Observable {@link List} of {@link AccountRestrictions}
+     */
+    Observable<List<AccountRestrictions>> getAccountsRestrictionsInfoFromPublicKeys(List<PublicKey> publicKeys);
```
Comment (rg911) [inline sdk-core/src/main/java/io/nem/sdk/api/AccountRepository.java]: Don't need this method. Only one method `getAccountsRestrictions` with 'List<Address>' as the param should be enough as Address can be created derived from PublicKey. So only `{ Addresses: []}` to feed the rest.

File: sdk-core/src/main/java/io/nem/catapult/builders/AccountRestrictionTypeDto.java [hunk 2]
```diff
@@ -33,7 +33,20 @@
...
-    BLOCK((byte) 128);
+    BLOCK((byte) 128),
+
+    /**
+     * Catapult generated code patch!!
+     */
+    ALLOW_OUTGOING_ADDRESS((byte) 65),
+    ALLOW_OUTGOING_TRANSACTION_TYPE((byte) 68),
+    SENTINEL((byte) 5),
+    BLOCK_ADDRESS((byte) 129),
```
Comment (rg911) [inline sdk-core/src/main/java/io/nem/catapult/builders/AccountRestrictionTypeDto.java]: If the we use full restriction Type here. (BLOCK_OUTGOING.....) we should change the e.g `ADDRESS` to `ALLOW_INCOMING_ADDRESS` to make the enum key consistent.

File: sdk-core/src/main/java/io/nem/sdk/api/AccountRepository.java (hunk 1 above)
Comment (fboucquez) [reply sdk-core/src/main/java/io/nem/sdk/api/AccountRepository.java]: Done

File: sdk-core/src/main/java/io/nem/catapult/builders/AccountRestrictionTypeDto.java (hunk 2 above)
Comment (fboucquez) [reply sdk-core/src/main/java/io/nem/catapult/builders/AccountRestrictionTypeDto.java]: I didn't want to modify the generated code too much. The enum is private to the sdk, the clients should use the model object. Should I modify it anyway?

**Pull request merged**